### PR TITLE
Make management port available outside of the pod

### PR DESCRIPTION
**Pull request by tomashibm**

### CODE REVIEW

File: pkg/varnishcluster/controller/varnishcluster_statefulset_test.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ var _ = Describe("statefulset", func() {
 			Expect(varnishContainer.Args).To(Equal([]string{
 				"-F",
 				"-S", "/etc/varnish-secret/secret",
-				"-T", "127.0.0.1:6082",
+				"-T", "0.0.0.0:6082",
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_statefulset_test.go:104]: Can you explain in more detail how changing the address from `127.0.0.1` to `0.0.0.0` exposes the management port in a secure fashion? I understand that `127.0.0.1` (localhost) is only accessible from inside the pod, but doesn't this change allow anyone to hit port 6082?

**Review by kscharm — COMMENTED** [review]
Need some more clarification for why this change is necessary.

File: pkg/varnishcluster/controller/varnishcluster_statefulset_test.go (hunk 1 above)
Comment (tomashibm) [reply pkg/varnishcluster/controller/varnishcluster_statefulset_test.go:104]: Yes, it allows anyone inside the same network to access that management port, but it is still protected by the authentication by secret (the `"-S", "/etc/varnish-secret/secret"` parameter). The client should know it before establishing the connection.

**Review by cin — COMMENTED** [review]
Add logic to create a service based on a flag? Error out if that flag is set and sharding is enabled? Or just let it start knowing that the service won't broadcast to all pods?

We could also include something like [this](https://github.ibm.com/TheWeatherCompany/icm-cassandra/blob/feature/536-ingress-update/utils/kpe) -- executes the command on all pods that match a label, in parallel.

**Comment by tomashibm** [comment]
I'm still not sure if it is a viable solution with service. It will only correctly and consistently work when we have one pod behind that service. With two and more pods (doesn't matter if sharded or round robin) you can't know anymore which pod will receive the request. And since Varnish doesn't coordinate such requests, it doesn't seem very helpful to me.

Also we can't determine if sharding is enable since it's not a flag, but a specific VCL code. 

We can deploy it "as is" with a note in the docs that it won't broadcast, but so far I don't see a use case where it could be acceptable except the single node cluster case.

I see two solutions so far:

1. The users need to find out the hosts (pod IPs) by themselves and iterate over the nodes. After PR it will be possible to reach the control port and do it. Also the user could build a VCL that get the list of Varnish hosts. It is possible now through templating we provide.
2. Execute the command from the operator on all nodes like you suggest. Need to do some designing on this first, not sure how it should look like.

**Comment by cin** [comment]
I forgot about sharding not being known in config...and you're right the service has too many limitations. If we went with solution 2 above, would this have its own CR and reconcile live in the operator? Or should we do something more general with shell scripting/reuse what's in icm-cassandra?

**Comment by tomashibm** [comment]
With shell scripting it can be implemented even today outside the operator. I don't think it needs any support from the operator.

The other way, by implementing it in the operator, I think it would be implemented in Go code. Either the operator do the connection to pods and execute the commands, or the sidecar will receive a command somehow and do the `varnishadm` commands.

As for the interface not so sure. A separate CRD sounds okay to me. Something that user can deploy with an automation or manually and then see the status of execution. We can also clean up them with a configurable ttl since they are not really needed to live long term. Sounds kind of like a job, but without the need to deploy a pod to execute the command.

The existing CRD doesn't sound good here since it's not a state of the cluster, but an imperative action. Not sure how we could express something like that in a declarative way, but open to suggestions.

**Comment by cin** [comment]
Agreed. The whole thing seems moot though if we don't provide a way to hit the port (at least for this use case) w/out port forwarding. The other option is the have a script exec into the pod and run the command, which we don't need to move the control interface in that case.

**Comment by cin** [comment]
Talked about this in our standup yesterday. Think we're going to avoid the service for now and provide an example of how to call the admin interface directly with a named host (possible since we use a statefulset -- thanks Tomash!). We may consider doing something in the operator in the future but it seems scripting it is easy enough for now.

**Review by cin — APPROVED** [review]
Can add usage doc later if desired...just to get this merged quickly

**Pull request merged**